src/replication.c

```diff
@@ -125,37 +127,330 @@ static int handle_mgmt_event_fd(replica_t *replica);
- *  r_factor = replication_factor
+ * Case 1: Rebuild from healthy replica
+ * Case 2: Mesh rebuild i.e all replicas are downgrade
  */
-void
-update_volstate(spec_t *spec)
+static int 
+send_prepare_for_rebuild(spec_t *spec, replica_t *target_replica,
+    replica_t *healthy_replica)
 {
-	uint64_t max;
+	int ret = 0;
+	int replica_cnt;
+	uint64_t size;
+	uint64_t num = 1;
 	replica_t *replica;
+	zvol_io_hdr_t *rmgmtio = NULL;
+	uint64_t data_len = strlen(spec->volname) + 1;
+	zvol_op_code_t mgmt_opcode = ZVOL_OPCODE_PREPARE_FOR_REBUILD;
+	struct rcommon_mgmt_cmd *rcomm_mgmt;
+	mgmt_cmd_t *mgmt_cmd;
+
+	/* Store target_replica ptr and mark rebuild_in_progress to true */
+	spec->target_replica = target_replica;
+	spec->rebuild_in_progress = true;
+
+	replica_cnt = count_of_replicas_helping_rebuild(spec, healthy_replica);
+	size = replica_cnt * sizeof (mgmt_ack_t); 
+	rcomm_mgmt = allocate_rcommon_mgmt_cmd(size);
+	// Case 1
+	if (healthy_replica != NULL) {
+
+		mgmt_cmd = malloc(sizeof (mgmt_cmd_t));
+		BUILD_REPLICA_MGMT_HDR(rmgmtio, mgmt_opcode, data_len);
+
+		mgmt_cmd->io_hdr = rmgmtio;
+		mgmt_cmd->io_bytes = 0;
+		mgmt_cmd->data = (char *)malloc(data_len);
+		snprintf((char *)mgmt_cmd->data, data_len, "%s", spec->volname);
+		mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+		mgmt_cmd->rcomm_mgmt = rcomm_mgmt;
+
+		MTX_LOCK(&healthy_replica->r_mtx);
+		TAILQ_INSERT_TAIL(&healthy_replica->mgmt_cmd_queue, mgmt_cmd,
+		    mgmt_cmd_next);
+		MTX_UNLOCK(&healthy_replica->r_mtx);
+
+		rcomm_mgmt->cmds_sent++;
+	
+		if (write(healthy_replica->mgmt_eventfd1, &num,
+		    sizeof (num)) != sizeof (num)) {
+			REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd "
+			    "for replica(%p)\n", replica);
+			ret = -1;
+			rcomm_mgmt->cmds_failed++;
+			goto exit;
+		}
+	
+		return ret;
+	}
+
+	// Case 2
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if (replica == target_replica)
+			continue;
+		
+		if (replica_cnt == 0)
+			break;
+		replica_cnt--;
+
+		mgmt_cmd = malloc(sizeof (mgmt_cmd_t));
+		BUILD_REPLICA_MGMT_HDR(rmgmtio, mgmt_opcode, data_len);
+
+		mgmt_cmd->io_hdr = rmgmtio;
+		mgmt_cmd->io_bytes = 0;
+		mgmt_cmd->data = (char *)malloc(data_len);
+		snprintf((char *)mgmt_cmd->data, data_len, "%s", spec->volname);
+		mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+		mgmt_cmd->rcomm_mgmt = rcomm_mgmt;
+	
+		MTX_LOCK(&replica->r_mtx);
+		TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd,
+		    mgmt_cmd_next);
+		MTX_UNLOCK(&replica->r_mtx);
+		rcomm_mgmt->cmds_sent++;
+
+		if (write(replica->mgmt_eventfd1, &num, sizeof (num)) !=
+		    sizeof (num)) {
+			REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd "
+			    "for replica(%p)\n", replica);
+			ret = -1;
+			rcomm_mgmt->cmds_failed++;
+			goto exit;
+		}
+	}
+
+exit:
+	if (ret == -1) {
+		MTX_LOCK(&replica->r_mtx);
+		TAILQ_REMOVE(&replica->mgmt_cmd_queue, mgmt_cmd,
+		    mgmt_cmd_next);
+		MTX_UNLOCK(&replica->r_mtx);
+		if (rcomm_mgmt->cmds_failed == rcomm_mgmt->cmds_sent) {
```

Comment:
lets also add comment that why we don't need lock of rcomm_mgmt to access these variables